### ISSUE TITLE
Should Image API v3 allow upscaling?

### ISSUE THREAD

**Issue by zimeon** [IIIF/api#1627]
Should the Image API v3 allow upscaling of image size as Image API v2 does? Two use cases where this might be useful/better than simply doing it in the browser are:

  1. when a server might use different/better algorithm for upscaling than a browser (note however, there is now some CSS control over browser behavior https://css-tricks.com/almanac/properties/i/image-rendering/)
  2. when a server actually has more information that is exposed via the IIIF Image API service (e.g. the underlying image content might be at higher resolution that the `width` and `height` made available via the Image API, and this data is used in upscaling requests)

Earlier discussions include:
  * Meeting notes from earlier discussion https://docs.google.com/document/d/1WysZu6e94zgyr9vX5rgd1qnFxLhEWeWqjLTIKivGbhE/edit
  * Revisits issues #693, #1370, #1372

(Issue was raised by @beaudet at the IIIF Washington Conference 2018)

**Comment by zimeon** [comment]
Discussion on [2018-01-31 call](https://docs.google.com/document/d/1WysZu6e94zgyr9vX5rgd1qnFxLhEWeWqjLTIKivGbhE/edit) has notes on #693/#678:
> Does anyone really need upscaled images, where max>full?
> Browsers are now as good as servers at upscaling
> Deprecate “size above full” seems easiest to implement and avoids semantic conflict. Many +1’s in > chat. This is the general consensus of the call, note will be placed in the ticket.

Editors discussion from 2018-06-13: This issue was discussed previously and determined insufficiently significant to warrant support. Changing the Image 3.0 specification to again support upscaling would have knock-on effects that would undo a number of simplifications in the new specification. We don’t believe that this is justified but seek other opinions.

**Comment by zimeon** [comment]
@beaudet - have I captured the issue accurately? any additional comments?

**Comment by beaudet** [comment]
Yes, you have accurately captured the issue.  Browsers are OK at upscaling (better than downscaling at least), but there are two problems with it.

1. This statement seems to be based on an underlying assumption that the IIIF image API is limited to providing images to web browsers.  I think that assumption artificially and tremendously limits the applications to which the Image API can be used.  Relying on the capabilities of a specific technology, in this case web browsers, is something that the editors have generally tried to avoid.  

2. This statement is false.  On the surface this statement seems easy to accept because browsers have improved quite a bit over the years at upscaling, but they still do NOT upscale as well as some IIIF image servers.

For example, using the same source pixels (296 x 210 source image), the latest Chrome web browser, using a fixed width="768" and height="545" attribute, produces this image:
![image](https://user-images.githubusercontent.com/77699/41416237-65e8ce8a-6fb8-11e8-8ab7-453a3cf55f40.png)

whereas IIP, using exactly the same source pixels and upscaling to the same size produces this image:
![image](https://user-images.githubusercontent.com/77699/41416283-7db76ed6-6fb8-11e8-9aa3-9481d5923707.png)

Open these images in two tabs in your browser and swap back and forth.  You'll see the upscaling is quite a bit better from IIP than from the browser.  Pay particular attention to areas of contrast like around the straw and the baked edge of the cake.  The IIP image is, at least to me, clearly superior.  If I was designing a web site and wanted all my images to be rendered at the same resolution and didn't have quite the right number of pixels in all my sources, I would surely want to use the IIP resized images rather than relying on browser.  This approach also removes the likely variability in web browser resizing quality.

So, in summary, I'm fine with changing the default behavior as it's not really a breaking change since upscaling has always been optional but I still think there should be a syntax, perhaps a new one, available via the Image API to return an upscaled image when an image server supports upscaling.  Eliminating the possibility of upscaling entirely would represent a step backwards in image quality when presenting through a web browser and in non-web browser cases, upscaling would have to be reimplemented by every client.  Non web browser cases might include fetching images for the purpose of analysis or a command line script that's composing a huge image from a series of IIIF images, code that is producing PDFs with embedded images, etc.

**Comment by azaroth42** [comment]
Proposal to allow `sizeAboveFull` in 3.0 and keeping the intent of the solutions for #693, #1370 and #1372:

* Servers MUST advertise the capability in `extraFeatures` of info.json. It is optional at level 2.
* Leave the feature named as `sizeAboveFull` despite ...
* Leave `full` size as removed
* Clarify that `max` is the largest image that can be returned by the server that is less than or equal to the `height` and `width` of full as advertised in info.json
* Clarify that `!h,w` syntax is limited to be less than or equal to `max`
* Clarify that there is no way to advertise the `max-max` (e.g. you can't say that 195% is okay, but 201% is not)

Tristan's use case of wanting an image that fits within an 800x800 box without upscaling is solved by the second to last bullet, at the expense of the inconsistency of ! having different rules than the other parameters.  If you really want an image that is arbitrarily tall, and 800 wide, then !800,100000000 would generate it.  It could also be coerced into something more sensible by /square/!800,800/ rather than /full/!800,800/

**Comment by beaudet** [comment]
So a client would have to read info.json and make the calculations for the desired size before making a request to retrieve pixels beyond the image limits?  If so, I'm not a huge fan of that.  While I'm fine with clarifying the existing behavior, I think it would be more useful and simpler for developers / designers if they could still request an aspect ratio honored bounding box larger than the image and receive an upscaled image if the server supports it.  Maybe something like "!!h,w" ?

**Comment by zimeon** [comment]
In the above prescription I wonder whether renaming `sizeAboveFull` to `sizeUpscaling` or such would make it a bit less confusing

**Comment by beaudet** [comment]
Ok, so to clarify, sizeAboveFull replaces "full" in API 3 and is not just a "feature" listed in info.json?  In other words, the following URL would allow an upscaled image to be produced from a bounding box? /sizeAboveFull/!w,h/... ?   If so, that's great.  I agree with @zimeon that the name of the feature could benefit from tweaking.

**Comment by tomcrane** [comment]
+1 to a name that focuses on _upsampling_ as the feature being enabled, rather than anything to do with the size of the response, and that the other suite of features all interpret max in relation to the advertised size.

**Comment by zimeon** [comment]
@beaudet - re. https://github.com/IIIF/api/issues/1627#issuecomment-400302945 - I am not suggesting `sizeAboveFull` or `sizeUpscaling` or `sizeUpsampling` as a region parameter. We are talking about a feature name (cf. `sizeAboveFull` in 2.1) that says a server can support requests for upscaled/upsampled images (whether from the full region or a portion). @azaroth42's https://github.com/IIIF/api/issues/1627#issuecomment-400098884 is an attempt to work out how to reconcile the rules/description with the rest of 3.0.

**Comment by beaudet** [comment]
Ok, so my original question still applies I think - how can a request be made for a bounding box that allows upscaling when the bounding box is larger than the pixels currently available to the client?

**Comment by zimeon** [comment]
I think a key thing about size parameters is that all forms except _`w,h`_ are only really useful in the case that the client has not read `info.json`. If the client _has_ read `info.json` then it can work out what what it wants and specify _`w,h`_. So, the question then becomes, how does one specify rules for interpretation of these parameters that makes sense and is useful with and without server support for upscaling. _`w,`_, _`,h`_ and _`pct:n`_ all seem straightforward as currently defined (modulo their resulting in bigger image vs. HTTP 400 when upscaling is requested). I think that if we wanted to allow upscaling the following definitions for _`max`_ and _`!w,h`_ are plausible:

_`max`_  - The extracted region is returned at the maximum size available but not larger than the pixel dimensions of the extracted region. The resulting image might be constrained to a smaller size by `maxWidth`, `maxHeight`, or `maxArea` as defined in the Technical Properties section.

_`!w,h`_ - The extracted region is scaled so that the width and height of the returned image are not greater than _`w`_ and _`h`_, while maintaining the aspect ratio. The returned image must be as large as possible but not larger than _`w`_ or _`h`_, or server-imposed limits. If the `sizeUpscaling` feature is supported then the returned image may be larger than the pixel dimensions of the extracted region, otherwise it _MUST_ not be larger than the pixel dimensions of the extracted region.

This would then allow `/full/!800,800/0/default.jpg` to mean give me an image that is as big as you are willing to give me that fits within an 800,800 box (and if the source were 100x100 it would be upscaled 8 fold it the server supported `sizeUpscaling`). This is different from @azaroth42's 5th bullet in https://github.com/IIIF/api/issues/1627#issuecomment-400098884 and does muddy the waters regarding the definition of `max`

**Comment by beaudet** [comment]
and that behavior would be backwards compatible with the 2.1 spec.  

Since there seem to be use cases for both upscaling and specifically not upscaling in responses, and since forced non-upscaling is arguably a new behavior, maybe the requirement just needs to be flipped around?  The default behavior of /full/!800,800/0/default.jpg could be retained (upscale if supported), but perhaps a new behavior defined by a syntax modification, e.g. "!!" (or something else), could indicate that no upscaling should take place even if it's supported.  This would let clients dial in the exact behavior they are looking for and it wouldn't change the expectations of existing clients.

**Comment by azaroth42** [comment]
Given that the accepted proposal was to get rid of the functionality completely, I think that `!w,h` not doing upscaling is a reasonable compromise. I'm not in favor of new syntax without at least one more institution speaking up in favor of an actual need for server side upscaling to be retrieved with the `!w,h` syntax.

**Comment by azaroth42** [comment]
Editors proposal, from call 2018-07-11:

On the community technical call, it was agreed that server side upscaling was a useful feature in some situations, and should not be removed completely from the 3.0 Image API (notably Dave Beaudet, NGA).  It was also agreed that the issue that prompted its removal was important to consider and keep in scope (notably Tristan Roddis, Cogapp). 

Proposal from the editors to try and find the appropriate solution that meets both:

Change the size parameter of the Image API to include an optional prefix of `^` (caret) before any of the existing parameter values.  If `^` is present, then a server that supports scaling the image beyond the dimensions published as its `height` and `width` (as applied to the extracted region), will return an upscaled image if possible and otherwise an error. If `^` is not present, then it MUST return an error if the requested size is greater than the extracted image's dimensions.

Servers that support upscaling MUST use the `maxHeight`, `maxWidth` and `maxArea` properties of info.json to publish the maximum size that they are willing to deliver any single derivative at, regardless of size of the extracted region.  Thus if the extracted region is 0,0,1,1 (a single pixel) and the server advertises a `maxWidth` of 10000, then it must deliver that single pixel scaled to 10000x. 

Support for upscaling is optional at all conformance levels, but the `^` syntax MUST be parsed and not be treated as a client syntax error.

For a 1000x1000 image, with the `full` region, the meaning of various size parameters:
* ^!10000,10000 = the largest image that fits within 10000,10000 including upscaling if upscaling is supported.  Thus the image could be anywhere from 1000,1000 up to 10000,10000.
* !10000,10000 = the largest image that fits within 10000,10000 without upscaling. Thus the image returned would be 1000,1000 (unless `max*` is less than 1000, at which point it would be that)
* ^10000,10000 = give me a 10000,10000 image, and if you have to upscale, then do so. Thus the image returned would be 10000,10000, or an error if the server does not support upscaling, or to that degree (10x)
* 10000,10000 = give me a 10000,10000 image, without upscaling.  Thus the server will return an error, as upscaling is not allowed and the size requested is greater than the maximum size.

A client could determine whether or not the requests will be successful by inspecting `max*` of the info.json, which would be 10000 or higher if the request is supported.

As a note, the `+` (plus) sign was considered instead of `^`, but if the server puts the IIIF parameters as a query parameter (like some versions of IIPImage Server do), then the + would be translated to a space, due to the encoding rules of www-url-form-encoded.

**Comment by beaudet** [comment]
Thanks for accommodating the use case in Image API v3.  This solution seems as clean as possible and I concur about the potential problems with using "+".  The syntax extension of "^" suggests "up" and so it's easy to remember as the "upscaling" syntax.   The examples you gave above will be helpful for the API v3 compatibility tool.

**Comment by azaroth42** [comment]
Closed by #1682! :)
